src/Crypto/EncryptionTrait.php

```diff
@@ -0,0 +1,177 @@
+        }
+        if (!is_int($cipherOptions['KeySize'])) {
+            throw new \InvalidArgumentException('The cipher "KeySize" must be'
+                . ' an integer.');
+        } elseif (!MaterialsProvider::isSupportedKeySize(
+            $cipherOptions['KeySize']
+        )) {
+            throw new \InvalidArgumentException('The cipher "KeySize" requested'
+                . ' is not supported by AES (128, 192, or 256).');
+        }
+
+        $cipherOptions['Iv'] = $provider->generateIv(
+            $this->getCipherOpenSslName(
+                $cipherOptions['Cipher'],
+                $cipherOptions['KeySize']
+            )
+        );
+
+        $cek = $provider->generateCek($cipherOptions['KeySize']);
+
+        list($encryptingStream, $aesName) = $this->getEncryptingStream(
+            $plaintext,
+            $cek,
+            $cipherOptions
+        );
+
+        // Populate envelope data
+        $envelope[MetadataEnvelope::CONTENT_KEY_V2_HEADER] =
+            $provider->encryptCek(
+                $cek,
+                $materialsDescription
+            );
+        unset($cek);
+
+        $envelope[MetadataEnvelope::IV_HEADER] =
+            base64_encode($cipherOptions['Iv']);
+        $envelope[MetadataEnvelope::KEY_WRAP_ALGORITHM_HEADER] =
+            $provider->getWrapAlgorithmName();
+        $envelope[MetadataEnvelope::CONTENT_CRYPTO_SCHEME_HEADER] = $aesName;
+        $envelope[MetadataEnvelope::UNENCRYPTED_CONTENT_LENGTH_HEADER] =
+            strlen($plaintext);
+        $envelope[MetadataEnvelope::UNENCRYPTED_CONTENT_MD5_HEADER] =
+            base64_encode(md5($plaintext));
+        $envelope[MetadataEnvelope::MATERIALS_DESCRIPTION_HEADER] =
+            json_encode($materialsDescription);
+        if (!empty($cipherOptions['Tag'])) {
+            $envelope[MetadataEnvelope::CRYPTO_TAG_LENGTH_HEADER] =
+                strlen($cipherOptions['Tag']) * 8;
+        }
+
+        return $encryptingStream;
+    }
+
+    /**
+     * Generates a stream that wraps the plaintext with the proper cipher and
+     * uses the content encryption key to encrypt the data when read.
+     *
+     * @param Stream $plaintext Plain-text data to be encrypted using the
+     *                          materials, algorithm, and data provided.
+     * @param string $cek A content encryption key for use by the stream for
+     *                    encrypting the plaintext data.
+     * @param array $cipherOptions Options for use in determining the cipher to
+     *                             be used for encrypting data.
+     *
+     * @return [AesStreamInterface, string]
+     *
+     * @internal
+     */
+    protected function getEncryptingStream(
+        Stream $plaintext,
+        $cek,
+        &$cipherOptions
+    ) {
+        switch ($cipherOptions['Cipher']) {
+            case 'gcm':
+                $cipherOptions['TagLength'] = 16;
+
+                $taggedStream = new AesTaggedGcmEncryptingStream(
```

Comment:
Couldn't `AesTaggedGcmEncryptingStream` be removed if you just used an [`AppendStream`](https://github.com/guzzle/psr7#appendstream) here?